src/main/flight/mixer.h

```diff
@@ -69,6 +69,7 @@ typedef struct mixerConfig_s {
     bool hasFlaps;
     int16_t appliedMixerPreset;
     uint16_t fwMinThrottleDownPitchAngle;
+    uint8_t pitchRollWeight;
```

Comment:
This doesn't appear to be used anywhere yet. What's the purpose of this setting?

Reply:
It is used in configurator. It's here to save these setting in controller flash between configurator launches. If I should do this other way, say how. I didn't find another way for now.

Reply:
If it's not going to be used in firmware - not sure if we should keep this setting here. What's the purpose of it?

Reply:
In helicopter CCPM consists of, for example, 3 servos on which the swashplate hangs.
![option-plateau-cyclique-alu-microheli-pour-balde-nano-cpx-](https://user-images.githubusercontent.com/50098090/57578499-5a468c00-7496-11e9-8568-4e3c325d2475.jpg)
All 3 servos can move in one direction which cause the swashplate to slide along the shaft - this called "collective pitch". One servo can move down, and other one move up which cause the swashplate to incline - this called "cyclic pitch" and used for pitch and roll control of helicopter.
So the mixer for servos looks like:
S1=ColPitch+Pitch * cos(0)+Roll * sin(0)
S2=ColPitch+Pitch * cos(120)+Roll * sin(120)
S3=ColPitch+Pitch * cos(240)+Roll * sin(240)
But it's mechanicaly impossible to incline the swashplate on such angle, that one servo if fully down and other is fully up. So we have to decrease pitch and roll weights in mixers for servo:
S1=ColPitch+(Pitch * cos(0)+Roll * sin(0)) * PitchRollWeight
S2=ColPitch+(Pitch * cos(120)+Roll * sin(120)) * PitchRollWeight
S3=ColPitch+(Pitch * cos(240)+Roll * sin(240)) * PitchRollWeight
On different helicopters this "PitchRollWeight" would vary widely - it mechanicaly defined. So when I write mixers for helicopter in inav-configurator I decide to create an option PitchRollWeight, which is used by configurator for mixer coefficients calculation. This option allows user to calculate mixer, load it into helicopter, move sticks, see what happened with the swashplate, change the setting and repeat - user don't have to change 5 mixer coeffs, but only one setting.
![photo_2019-05-12_08-55-24](https://user-images.githubusercontent.com/50098090/57578680-88799b00-7499-11e9-9afe-925672e6adfb.jpg)
I need to store this setting somewhere, for user can disconnect from computer, connect battery, move sticks or even commit a test flight, then connect to computer again and see his previous setting instead of the default one. I decided it would be good idea to store it onboard even if it is used only in configurator when calculating mixer coefficients. This allows user connect the controller to other computer and change mixer on it if he wants. So I decide to store this value in this structure cause I think it make sense. appliedMixerPreset, for example, is also here. It's also used only in configurator to make it more comfortable and not used in firmware.

Reply:
I see.

However, we encourage mixer rules in INAV to be configured consistently across the while family of the UAVs. I.e. for **all** flying wings we have mixer configured exactly the same way regardless of mechanical limits of the elevon throws.

The actual limit is placed on servo endpoints (min/mid/max), and mixer output is mapped to servo min/max endpoints to allow the servo to execute the whole range of mixer output.

I don't think we need a separate setting for this, especially the setting that will make mixer rules inconsistent across the family of aircrafts. Mixer should define core logic, so for CPPM120 heli it should define only this:

```
S1=ColPitch+Pitch * cos(0)+Roll * sin(0)
S2=ColPitch+Pitch * cos(120)+Roll * sin(120)
S3=ColPitch+Pitch * cos(240)+Roll * sin(240)
```

The actual servo throws should be configured with servo endpoints to keep things consistent.

If we want a simple multiplier on servo throes we should add that to the servo mixer. @DzikuVx what do you think?

Reply:
Actually I forgot that we already have servo rate 😆 
So we don't need a new parameter at all.

Reply:
Of course, the actual servo throws should be configured with servo endpoints. This parameter is not about servo rates.
For example in my helicopter (one of it) I have servos:
S1 - 1180-1850 us
S2 - 1220-1870 us
S3 - 1800 - 1150 us - it stands upside down
I set min/max values for these servos, set invert checkbox for for S3. That's great.
Let's see what's happened now in different cases (I suppose all values are in -1~1 range):
1. ColPitch=1, Pitch=0, Roll=0:
`S1=ColPitch+Pitch * cos(0)+Roll * sin(0) = 1 (1850 us)`
`S2=ColPitch+Pitch * cos(120)+Roll * sin(120) = 1 (1870 us)`
`S3=ColPitch+Pitch * cos(240)+Roll * sin(240) = 1 (1150 us)`
This position is possible - swashplate is in the upper position.
2. ColPitch=0, Pitch=1, Roll=0:
`S1=ColPitch+Pitch * cos(0)+Roll * sin(0) = 1 (1850 us)`
`S2=ColPitch+Pitch * cos(120)+Roll * sin(120) = -0.5 (1383 us)`
`S3=ColPitch+Pitch * cos(240)+Roll * sin(240) = -0.5 (1638  us)`
This position is mechanically impossible - swashplate can not incline on such angle. The servos will lock and break something or themselves.
3. ColPitch=-0.9, Pitch=0.2, Roll=0:
`S1=ColPitch+Pitch * cos(0)+Roll * sin(0) = -0.7 (1281 us)`
`S2=ColPitch+Pitch * cos(120)+Roll * sin(120) = -1 (1220 us)`
`S3=ColPitch+Pitch * cos(240)+Roll * sin(240) = -1 (1800  us)`
This position is possible again - swashplate is in the lower position and inclined a little.
So it's impossible to limit servo min/max values to avoid impossible positions. We have to limit Pitch and Roll values till [-0.2~0.2] range for example. As far as I know, it's impossible to set limits for Pitch and Roll values (not servos) in iNav.
So I think like that:
We can't limit Pitch and Roll values to [-0.2~0.2], but we can rewrite formula like:
`S1=ColPitch+(Pitch * PitchWeight) * cos(0)+(Roll * RollWeight) * sin(0)`
In that case ColPitch, Pitch and Roll are in [-1~1] range - everything is OK.
Then this formula rewrites:
`S1=ColPitch+Pitch * (PitchWeight * cos(0))+Roll * (RollWeight * sin(0))`
So we have mixer coefficients:
`1; PitchWeight * cos(0); RollWeight * sin(0)`
We really don't need separate coefficients for Pitch and Roll, so:
`1; PitchRollWeight * cos(0); PitchRollWeight * sin(0)`
I could set this values on my mind, like PitchRollWeight = 0.3 and write  mixer coefficients:
`1; 0.3*cos(0); 0.3*sin(0)`
But I have three helicopters and this parameter for them varies very much (all of them have the same CCPM): something like 0.2, 0.3 and 0.6 - this depends on swashplate construction (sizes). So I decide to make this parameter adjustable by user.

Reply:
I think PitchRollWeight doesn't break any mixer logic. The helicopter CCPM is the only aircraft part I know where we cannot avoid impossible positions using servo limits but have to use Pitch/Roll limits. So, as far I can understand we have to either write Pitch/Roll/Yaw and all other limits logic (which will be applicable and non-excessive only for helicopter CCPM) or use such coefficient in mixer tab. I think, the second way is much easier for developers and for users.

Reply:
This is identical to adjusting the servo rates for all 3 swash-plate servos at once, but this can be done without messing up with mixer weights. 

We should NOT use firmware config as a storage for configurator-side parameters (with a few exceptions). Think about the case when user will change this parameter and clicking "Save" without clicking "Apply mixer" first. The user will be under the impression that they changed something, while in reality they didn't.

We should instruct the user how to make use of existing settings, rather than inventing new ones.

Reply:
Remove pitchRollWeight from firmware. Move this option to configurator and create a helicopter mixer options modal dialog according to discussion with @digitalentity in Telegram.